lib/services/app.go

```diff
@@ -158,3 +169,217 @@ func UnmarshalAppServer(data []byte, opts ...MarshalOption) (types.AppServer, er
 	}
 	return nil, trace.BadParameter("unsupported app server resource version %q", h.Version)
 }
+
+// NewApplicationsFromKubeService creates application resources from kubernetes service. One service result
+// in multiple application resources if there are multiple ports exposed.
+func NewApplicationsFromKubeService(service v1.Service, clusterName string, pi ProtocolChecker) ([]types.Application, error) {
+	var apps types.Apps
+	protocol := service.GetAnnotations()[types.DiscoveryProtocolLabel]
+
+	ports, err := getServicePorts(service)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var appsMu sync.Mutex
+	g := errgroup.Group{}
+	g.SetLimit(5)
+	for _, port := range ports {
+		port := port
+		g.Go(func() error {
+			appURI := getAppURI(getServiceFQDN(service), port, protocol, pi)
+
+			// By default we are looking for http apps, we only create tcp apps if protocol annotation
+			// explicitly specifies it.
+			if strings.HasPrefix(appURI, protoTCP) && protocol != protoTCP {
+				return nil
+			}
+
+			rewriteConfig, err := getAppRewriteConfig(service.GetAnnotations())
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			a, err := types.NewAppV3(types.Metadata{
+				Name:        getAppName(service.GetName(), service.GetNamespace(), clusterName, ""),
+				Description: fmt.Sprintf("Discovered application in Kubernetes cluster %q", clusterName),
+				Labels:      getAppLabels(service.GetLabels(), clusterName),
+			}, types.AppSpecV3{
+				URI: appURI,
+				// Temporary usage to have app's name with port name, in case there are more than one app
+				// created from this service.
+				PublicAddr: getAppName(service.GetName(), service.GetNamespace(), clusterName, port.Name),
+				Rewrite:    rewriteConfig,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			appsMu.Lock()
+			apps = append(apps, a)
+			appsMu.Unlock()
+			return nil
+		})
+	}
+	err = g.Wait()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// If we ended up with more than one app from the service, we'll have port name in the app name to distinguish them.
+	// We can only do this now because we don't know how many apps there will be until all ports are processed by errgroup.
+	for i := range apps {
```

Comment:
I don't think this is necessary. If `len(ports)>0`, we can automatically do this above even if some ports can fail as it's easier to detect errors.

Can we also log the discovery errors?

I also feel that this automatic protocol detection should live somewhere under `srv/discovery`.

You can still have the `NewApplicationsFromKubeService` here, but it only receives the service with the protocol and a single port - failing on every other case. Moving every other logic to the discovery folder would allow simplifying the logic of where and how we build the services even if we must change the annotations and ports of the service we receive before calling the  `NewApplicationsFromKubeService` function. What do you think about it?

Reply:
As discussed offline I think that it's worth a bit of code to try to keep app names shorter for users.

I feel that protocol detection can be considered as a part of turning service into apps. We have pretty clean separation when discovery find services and then we handle transforming them here as a whole. If there would be another consumer (although I don't see candidates) it could reuse same logic.
So, I guess we could move it, but I would prefer not to.

Reply:
Yep, I selected those numbers for parallelism because I think 50 HEAD requests is ok limit for worst case scenario (which probably will not happen, since services usually won't have 5 ports requiring http pinging). We can lower parallel ports processing if we want or even remove parallelism from there - it's probably nice to have, but no that important. Important one is on the services level.

Reply:
Moved this code to discovery, as discussed offline